### PR TITLE
thunderscope: protobuf, g3log, networking and basic field widget with layers

### LINKED ISSUE #1974: Log robot and ball velocities from SensorFusion
### Description of the task

Currently, we convert `World` to `SSL_WrapperPacket` when we log it with a `ProtoLogger`. However, `SSL_WrapperPacket` does not have fields for robot or ball velocities. We need a way to log robot velocities from SensorFusion. 

There are two ways we could do this: 
1. Define a Protobuf message containing an `SSL_WrapperPacket` as well as repeated fields containing the additional velocity data, e.g. a map of robot id's to a vector representing its velocity
     - Pros: code reuse for both logging and plotting, easier to implement
     - Cons: storing velocities separately from other robot data is not ideal design
2. Define an entirely new message to store all the fields from `World`
     - Pros: the ideal design solution
     - Cons: a lot of effort, cannot reuse plotting tools

### Acceptance criteria

- [ ] Log robot and ball velocities from SensorFusion
- [ ] Unit tests for the function that does the conversion

### LINKED ISSUE #2274: Unix Socket Protobuf Communication Python Scripts
### Description of the task
We'd like to experiment with Unix Socket Communication, so the goal here is to make two Python scripts to send protobufs back and forth over Unix sockets

### Acceptance criteria

- [ ] two python scripts to send protobufs back and forth over Unix sockets

### LINKED ISSUE #2300: Define Path and Obstacle protobuf
### Description of the task
Add a Circle and Polygon message to geometry.proto, then use geometry.proto to define a Path (repeated `Point`s) and Obstacle (oneof Polygon and Circle)

### Acceptance criteria

- [ ] Add a Circle and Polygon message to geometry.proto
- [ ] use geometry.proto to define a Path (repeated `Point`s) and Obstacle (oneof Polygon and Circle)
- [ ] conversion functions from `ObstaclePtr` and `std::vector<Point>`

### PR DESCRIPTION
**Pull request by akhilveeraghanta**
### Description

First PR for thunderscope! Lot more work to come

* Base64 Library
* Python Threaded Unix Listener
* C++ Threaded Unix Sender
* g3log Integration
* Thunderscope Main
* remove hermetic python (it doesn't play nicely with qt unfortunately) and add a tbots venv in /opt

### Testing Done
Manual testing

bazel run //software/thunderscope:thunderscope
### Resolved Issues
resolves #1974
resolves #2274
resolves #2300 

### Length Justification
most of the changes are build files, documentation and trivial changes (at least half)

the unix socket code + networking + plotting code is the bulk of this PR

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/thunderscope/field/field_layer.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        }
+
+    def boundingRect(self):
+        """boundingRect _must_ indicate the entire area that will be drawn on or
+        else we will get artifacts and possibly crashing.
+    
+        :return: Rectangle that covers the entire field
+    
+        """
+        # TODO (#TODO) this rectangle makes no sense, it should be
```
Comment (akhilveeraghanta) [inline src/software/thunderscope/field/field_layer.py]: Note to reviewer: I am making tickets for all the #TODOs in parallel, there are quite a few, don't worry about them for the first review pass

File: src/software/thunderscope/field/obstacle_layer.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+        self.cached_obstacles = obstacles
+
+        painter.setPen(pg.mkPen(colors.NAVIGATOR_OBSTACLE_COLOR))
+
+        for polyobstacle in obstacles.polygon:
+            polygon_points = [
+                QtCore.QPoint(
+                    constants.MM_TO_M * point.x_meters,
+                    constants.MM_TO_M * point.y_meters,
```
Comment (tarakong01) [inline src/software/thunderscope/field/obstacle_layer.py]: I'm pretty sure this is converting m to mm, but the name of the constant MM_TO_M is a bit confusing since it sounds like it's converting from mm to m, could we rename it to M_TO_MM or MM_PER_M

**Review by tarakong01 — DISMISSED** [review]
Just one minor comment but this looks awesome, I'm excited to get back to working on this!

File: src/proto/visualization.proto [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+{
+    repeated Point point = 1;
+}
+
+message PathVisualization
+{
+    repeated Path path = 1;
+}
+
+message Obstacle
```
Comment (jonathanlew) [inline src/proto/visualization.proto]: nit: this represents multiple Obstacles so perhaps rename to Obstacles

File: src/software/gui/drawing/obstacle_artist.cpp [hunk 4]
```diff
@@ -1,21 +1,53 @@
 #include "software/gui/drawing/obstacle_artist.h"
 
+#include "proto/geometry.pb.h"
+#include "proto/message_translation/tbots_geometry.h"
+#include "software/logger/logger.h"
+
 ObstacleArtist::ObstacleArtist(QGraphicsScene* scene, const QPen& pen)
     : scene_(scene), pen_(pen)
```
Comment (jonathanlew) [inline src/software/gui/drawing/obstacle_artist.cpp]: nit: explicitly initialize obstacle_proto_

File: src/software/logger/protobuf_sink.cpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    else
+    {
+        TbotsProto::RobotLog log_msg_proto;
+        TbotsProto::LogLevel log_level_proto;
+
+        if (TbotsProto::LogLevel_Parse(log_entry.get().level(), &log_level_proto))
+        {
+            std::time_t t = std::time(0);
+            log_msg_proto.mutable_created_timestamp()->set_epoch_timestamp_seconds(
+                static_cast<double>(t));
```
Comment (jonathanlew) [inline src/software/logger/protobuf_sink.cpp]: why don't we use `std::chrono::time_point<std::chrono::steady_clock>` here?

File: src/software/logger/protobuf_sink.h [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+#include <fstream>
+#include <g3log/logmessage.hpp>
+#include <iostream>
+
+#include "google/protobuf/any.pb.h"
+#include "software/logger/custom_logging_levels.h"
+#include "software/networking/threaded_unix_sender.h"
+
+static std::string TYPE_DELIMITER = "!!!";
+static std::string UNIX_BASE_PATH = "/tmp/tbots/";
```
Comment (jonathanlew) [inline src/software/logger/protobuf_sink.h]: should these be part of the class?

File: src/software/logger/protobuf_sink.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+   private:
+    UnixSenderMap unix_senders;
+};
+
+/*
+ * Serialize a protobuf and pack it into a msg
+ *
+ * @param os The output stream
+ * @param message The message to serialize
+ */
```
Comment (jonathanlew) [inline src/software/logger/protobuf_sink.h]: nit: `@return The output stream...`

File: src/software/networking/BUILD [hunk 8]
```diff
@@ -30,14 +32,13 @@ cc_library(
...
     deps = [
-        "//software/logger",
         "@boost//:asio",
     ],
 )
 
 cc_library(
     name = "threaded_proto_udp_listener",
-    hdrs = [
+    srcs = [
```
Comment (jonathanlew) [inline src/software/networking/BUILD]: nit: why this change?

File: src/software/networking/threaded_unix_sender.h [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+class ThreadedUnixSender
+{
+   public:
+    /**
+     * Creates a UnixSender that sends a string over the unix socket path.
+     *
+     * @param io_service The io_service to use to service outgoing SendProto data
+     * @param unix_socket_path The path to the unix socket
+     */
+    ThreadedUnixSender(const std::string& unix_socket_path);
```
Comment (jonathanlew) [inline src/software/networking/threaded_unix_sender.h]: nit: there's only one argument in this function

File: src/software/networking/threaded_unix_listener.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            any_msg.Unpack(msg)
+        else:
+            msg = self.proto_type.FromString(p)
+
+        self.handle_callback(msg)
+
+
+def handler_factory(proto_type, handle_callback, convert_from_any):
+    def create_handler(*args, **keys):
+        return Session(proto_type, handle_callback, convert_from_any, *args, **keys)
```
Comment (jonathanlew) [inline src/software/networking/threaded_unix_listener.py:128]: nit: docs for this function

File: src/software/networking/unix_sender.cpp [hunk 11]
```diff
@@ -0,0 +1,47 @@
+#include "software/networking/unix_sender.h"
+
+#include <iostream>
+
+static unsigned MAX_SEND_FAILURES_BEFORE_LOG = 1000;
```
Comment (jonathanlew) [inline src/software/networking/unix_sender.cpp]: nit: shouldn't this be in the header file?

File: src/software/thunderscope/field/path_layer.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+from pyqtgraph.Qt import QtCore, QtGui
+from proto.visualization_pb2 import PathVisualization
+from software.networking.threaded_unix_listener import ThreadedUnixListener
+
+
+class PathLayer(FieldLayer):
+    def __init__(self):
+        FieldLayer.__init__(self)
+        self.path_receiver = ThreadedUnixListener(
+            "/tmp/tbots/TbotsProto.PathVisualization", PathVisualization,
```
Comment (jonathanlew) [inline src/software/thunderscope/field/path_layer.py]: nit: I'd prefer if this weren't hardcoded

File: src/software/thunderscope/field/world_layer.py [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+from software.thunderscope.field.constants import ROBOT_MAX_RADIUS, BALL_RADIUS, MM_TO_M
+from software.networking.threaded_unix_listener import ThreadedUnixListener
+import software.thunderscope.colors as colors
+
+
+class WorldLayer(FieldLayer):
+    def __init__(self):
+        FieldLayer.__init__(self)
+        self.world_receiver = ThreadedUnixListener(
+            "/tmp/tbots/TbotsProto.World", World, max_buffer_size=1
```
Comment (jonathanlew) [inline src/software/thunderscope/field/world_layer.py]: perhaps /tmp/tbots/ should a constant?

File: src/software/thunderscope/field/world_layer.py [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+            self.createCircle(
+                ball.current_state.global_position.x_meters * MM_TO_M,
+                ball.current_state.global_position.y_meters * MM_TO_M,
+                BALL_RADIUS,
+            )
+        )
+
+    def paint(self, painter, option, widget):
+
+        world = self.world_receiver.maybe_pop()
```
Comment (jonathanlew) [inline src/software/thunderscope/field/world_layer.py]: nit: missing comments

File: src/software/thunderscope/log/g3log_widget.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+            """QPlainTextEdit{
+                color: #ffffff;
+                background-color: #000000;
+                selection-background-color: #606060;
+                selection-color: #ffffff;
+            }"""
+        )
+
+    def refresh(self):
+        log = self.log_receiver.maybe_pop()
```
Comment (jonathanlew) [inline src/software/thunderscope/log/g3log_widget.py]: nit: missing function comment

File: src/software/thunderscope/thunderscope.py [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    field.add_layer("Obstacles", obstacle_layer.ObstacleLayer())
+    field.add_layer("Path", path_layer.PathLayer())
+
+    field_dock = Dock("Field", size=(500, 2000))
+    field_dock.addWidget(field)
+
+    # Setup Console Widget
+    logs = g3logWidget()
+
+    log_dock = Dock("logs", size=(500, 100))
```
Comment (jonathanlew) [inline src/software/thunderscope/thunderscope.py:41]: would these sizes scale nicely to different displays?

File: src/software/networking/threaded_unix_listener.py [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+        """Buffer the protobuf, and raise a warning if we overrun the buffer
+
+        :param proto: The protobuf to buffer
+        :raises: Warning
+
+        """
+        try:
+            self.proto_buffer.put_nowait(proto)
+        except queue.Full as queue_full:
+            print("buffer overrun for {}".format(self.unix_path))
```
Comment (jonathanlew) [inline src/software/networking/threaded_unix_listener.py:69]: I'm getting a bunch of these running locally, is that expected? It seems kind of spammy

File: src/software/networking/threaded_unix_listener.py [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+
+    def serve_till_stopped(self):
```
Comment (jonathanlew) [inline src/software/networking/threaded_unix_listener.py:71]: nit: comments for these functions

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Nice work! pyqtgraph docking is super nice to use :sparkles:  Just a bunch of minor comments

File: src/software/gui/drawing/obstacle_artist.cpp [hunk 19]
```diff
@@ -1,21 +1,53 @@
...
 }
 
 void ObstacleArtist::visit(const GeomObstacle<Polygon>& geom_obstacle)
 {
+    TbotsProto::Polygon poly_proto;
+    std::vector<Point> points = geom_obstacle.getGeom().getPoints();
+    for (const auto& point : points)
+    {
+        *(poly_proto.add_points()) = *createPointProto(point);
+    }
```
Comment (nimazareian) [inline src/software/gui/drawing/obstacle_artist.cpp]: ~~We should probably add a `Polygon` -> `TbotsProto::Polygon` translation function to avoid duplicated translation code.~~ You should use `createPolygonProto` translation function here and below to create `poly_proto`. 
Also, a `Circle` translation function should probably be added and used above aswell to avoid duplicated code in the future.

File: src/software/thunderscope/field/obstacle_layer.py (hunk 2 above)
Comment (akhilveeraghanta) [reply src/software/thunderscope/field/obstacle_layer.py]: good catch, its definitely backwards, I changed it to MM_PER_M to be consistent with shared/constants.h

File: src/proto/visualization.proto (hunk 3 above)
Comment (akhilveeraghanta) [reply src/proto/visualization.proto]: changed

File: src/software/gui/drawing/obstacle_artist.cpp (hunk 4 above)
Comment (akhilveeraghanta) [reply src/software/gui/drawing/obstacle_artist.cpp]: done

File: src/software/logger/protobuf_sink.cpp (hunk 5 above)
Comment (akhilveeraghanta) [reply src/software/logger/protobuf_sink.cpp]: TIL about time_point, changed

File: src/software/logger/protobuf_sink.h (hunk 6 above)
Comment (akhilveeraghanta) [reply src/software/logger/protobuf_sink.h]: they can't be because the << operator needs them as well

File: src/software/logger/protobuf_sink.h (hunk 7 above)
Comment (akhilveeraghanta) [reply src/software/logger/protobuf_sink.h]: fixed

File: src/software/networking/BUILD (hunk 8 above)
Comment (akhilveeraghanta) [reply src/software/networking/BUILD]: was a find-replace mistake, fixed

File: src/software/networking/threaded_unix_sender.h (hunk 9 above)
Comment (akhilveeraghanta) [reply src/software/networking/threaded_unix_sender.h]: fixed

File: src/software/networking/threaded_unix_listener.py (hunk 10 above)
Comment (akhilveeraghanta) [reply src/software/networking/threaded_unix_listener.py:128]: added above

File: src/software/networking/unix_sender.cpp (hunk 11 above)
Comment (akhilveeraghanta) [reply src/software/networking/unix_sender.cpp]: moved to private member

File: src/software/thunderscope/field/path_layer.py (hunk 12 above)
Comment (akhilveeraghanta) [reply src/software/thunderscope/field/path_layer.py]: good point, we can generate this the same way protobufsink does, updated

File: src/software/thunderscope/field/world_layer.py (hunk 13 above)
Comment (akhilveeraghanta) [reply src/software/thunderscope/field/world_layer.py]: yup, pulled out

File: src/software/thunderscope/field/world_layer.py (hunk 14 above)
Comment (akhilveeraghanta) [reply src/software/thunderscope/field/world_layer.py]: added

File: src/software/thunderscope/log/g3log_widget.py (hunk 15 above)
Comment (akhilveeraghanta) [reply src/software/thunderscope/log/g3log_widget.py]: added

File: src/software/networking/threaded_unix_listener.py (hunk 18 above)
Comment (akhilveeraghanta) [reply src/software/networking/threaded_unix_listener.py:71]: added

File: src/software/gui/drawing/obstacle_artist.cpp (hunk 19 above)
Comment (akhilveeraghanta) [reply src/software/gui/drawing/obstacle_artist.cpp]: good catch! changed to use translation functions, looks nicer now

File: src/software/networking/threaded_unix_listener.py (hunk 17 above)
Comment (akhilveeraghanta) [reply src/software/networking/threaded_unix_listener.py:69]: you actually shouldn't be getting these _too_ much because it indicates that the plotting is falling behind "real time" and packets are being dropped

which msg buffer is it complaining about?

File: src/software/thunderscope/thunderscope.py (hunk 16 above)
Comment (akhilveeraghanta) [reply src/software/thunderscope/thunderscope.py:41]: im not sure actually, i've changed  my resolution and tried and it doesn't seem to change stuff from what i've gathered its a "size hint" maybe

@itsarune are you able to try this on your VM and see if it opens properly?

File: src/software/thunderscope/field/field_layer.py (hunk 1 above)
Comment (akhilveeraghanta) [reply src/software/thunderscope/field/field_layer.py]: done

File: src/proto/message_translation/tbots_geometry.cpp [hunk 20]
```diff
@@ -1,5 +1,7 @@
 #include "proto/message_translation/tbots_geometry.h"
 
+#include <google/protobuf/util/message_differencer.h>
```
Comment (nimazareian) [inline src/proto/message_translation/tbots_geometry.cpp]: Wondering if this is being used?

File: src/proto/message_translation/tbots_geometry_test.cpp [hunk 21]
```diff
@@ -52,3 +55,14 @@ TEST(TbotsProtobufTest, polygon_msg_test)
...
+
+TEST(TbotsProtobufTest, circle_msg_test)
+{
+    auto circle_1     = Circle(Point(-1, 1), 4);
+    auto circle_msg_1 = createCircleProto(circle_1);
+    auto circle_2     = createCircle(*circle_msg_1);
+    auto circle_msg_2 = createCircleProto(circle_2);
+
+    EXPECT_TRUE(
+        google::protobuf::util::MessageDifferencer::Equals(*circle_msg_1, *circle_msg_2));
```
Comment (nimazareian) [inline src/proto/message_translation/tbots_geometry_test.cpp:67]: Really interesting 🔥

**Review by nimazareian — COMMENTED** [review]
Great work, overall looks good to me. Just had two comments

File: src/proto/message_translation/tbots_geometry.cpp (hunk 20 above)
Comment (akhilveeraghanta) [reply src/proto/message_translation/tbots_geometry.cpp]: ah this was a mistake, removed!

File: src/software/thunderscope/thunderscope.py (hunk 16 above)
Comment (itsarune) [reply src/software/thunderscope/thunderscope.py:41]: Looks like the display is fine. I'll share a screenshot of what I see in my vm:
sc_VirtualBox_Lubuntu 18 04_15_01_2022_11_28_37

Comment (akhilveeraghanta) [reply src/software/thunderscope/thunderscope.py:41]: sweet thanks for running!, thats good, yeah then its just a size hint then, seems to scale it fine

File: src/proto/message_translation/tbots_geometry_test.cpp (hunk 21 above)
Comment (akhilveeraghanta) [reply src/proto/message_translation/tbots_geometry_test.cpp:67]: this isn't a perfect test because you could have:
~~~
circle_1 = Circle(1,1, radius=3)
circle_1_msg = createCircleProto(circle_1) // CircleMsg(1,1, radius=0) drops the radius value
circle_2 = createCircle(circle_1_msg) // Circle(1,1, radius=0) radius value is still 0
circle_2_msg = createCircleProto(circle_2) // CircleMsg(1,1, radius=0) drops the radius value

EXPECT_TRUE(circle_1_msg, circle_2_msg) is true but we've dropped the radius
~~~
probably worth checking if circle_1 == circle_2 ill add that too

File: src/software/networking/threaded_unix_listener.py (hunk 17 above)
Comment (jonathanlew) [reply src/software/networking/threaded_unix_listener.py:69]: mainly World and Obstacle and I guess it doesn't happen all the time so not blocking but something to keep in mind

**Review by jonathanlew — APPROVED** [review]
LGTM pending other reviewers' comments, nice work!

**Comment by jonathanlew** [comment]
It would be nice to add something to the getting-started guide

File: src/proto/message_translation/tbots_geometry_test.cpp (hunk 21 above)
Comment (akhilveeraghanta) [reply src/proto/message_translation/tbots_geometry_test.cpp:67]: I've made a ticket here to track this work for a later day:
https://github.com/UBC-Thunderbots/Software/issues/2403

**Review by nimazareian — APPROVED** [review]
LGTM!

File: src/proto/message_translation/tbots_geometry_test.cpp (hunk 21 above)
Comment (nimazareian) [reply src/proto/message_translation/tbots_geometry_test.cpp:67]: Aren't both radius and origin translated in both translation functions? Why would radius be dropped?
